docker/Dockerfile

```diff
@@ -51,40 +73,104 @@ RUN apt-get update && \
 	# Remove unzip and clean apt cache
 	apt-get clean && \
 	apt-get remove -y unzip && \
 	apt-get -y autoremove && \
 	rm -rf /var/lib/apt/lists/* 
 
-WORKDIR /opt/iotajson
+WORKDIR /opt/iotagent-json
 
+# hadolint ignore=DL3008
 RUN \
 	# Ensure that Git is installed prior to running npm install
-	apt-get update && \
-	apt-get install -y git && \
-	npm install pm2@3.2.2 -g && \
+	apt-get install -y --no-install-recommends git && \
 	echo "INFO: npm install --production..." && \
-	npm install --production && \
+	npm install --only=prod --no-package-lock --no-optional && \
 	# Remove Git and clean apt cache
 	apt-get clean && \
 	apt-get remove -y git && \
-	apt-get -y autoremove && \
-	chmod +x docker/entrypoint.sh
+	apt-get -y autoremove
 
 USER node
-ENV NODE_ENV=production
+ENV NODE_ENV=development
+# Expose 4041 for NORTH PORT, 7896 for HTTP PORT
+EXPOSE ${IOTA_NORTH_PORT:-4041} ${IOTA_HTTP_PORT:-7896}
+CMD ["node", "/opt/iotagent-json/bin/iotagent-json", "-- ", "config.js"]
+HEALTHCHECK --interval=30s --timeout=3s --start-period=10s \
+   CMD npm run healthcheck
 
-# Expose Ports - Default to 4041 for NORTH PORT 4041, 7896 for HTTP PORT
+########################################################################################
+#
+# This build stage installs PM2 if required.
+# 
+# To create an image using PM2 run this Docker file with the parameter:
+#
+# --target=pm2-install
+#
+########################################################################################
+FROM node:${NODE_VERSION} AS pm2-install
+
+COPY --from=builder /opt/iotagent-json /opt/iotagent-json
+RUN npm install pm2@4.4.0 -g --no-package-lock --no-optional
+
+USER node
+ENV NODE_ENV=development
+# Expose 4041 for NORTH PORT, 7896 for HTTP PORT
+EXPOSE ${IOTA_NORTH_PORT:-4041} ${IOTA_HTTP_PORT:-7896}
+CMD ["pm2-runtime", "/opt/iotagent-json/bin/iotagent-json", "-- ", "config.js"]
+
+########################################################################################
+#
+# This build stage creates an anonymous user to be used with the distroless build
+# as defined below.
+#
+########################################################################################
+FROM node:${NODE_VERSION} AS anon-user
+RUN sed -i -r "/^(root|nobody)/!d" /etc/passwd /etc/shadow /etc/group \
+    && sed -i -r 's#^(.*):[^:]*$#\1:/sbin/nologin#' /etc/passwd
+
+########################################################################################
+#
+# This build stage creates a distroless image for production.
+#
+# IMPORTANT: For production environments use Docker Secrets to protect values of the 
+# sensitive ENV variables defined below, by adding _FILE to the name of the relevant 
+# variable.
+#
+# - IOTA_AUTH_USER, IOTA_AUTH_PASSWORD - when using Keystone Security 
+# - IOTA_AUTH_CLIENT_ID, IOTA_AUTH_CLIENT_SECRET - when using OAuth2 Security
+#
+########################################################################################
+FROM gcr.io/distroless/nodejs:${NODE_VERSION}
```

Comment:
We know that this is a big step into security for containers, but we believe that since this container is still in construction, is not ready for a distroless upgrade. Regardless of that, there are plenty of users who might be affected for this change, since this project is open source.

Nevertheless we are not closed to this kinds of improvements to our containers, so we can accept this upgrade if this distroless container is a diferent target from the default one. The default one must be the usuall node-slim base image, and from the build hook. Both images could be created.

Reply:
Sounds like one solution could be three Dockerfiles - One for Node-slim, one for PM2 and one for Distroless.

Reply:
Do you mean different targets on the dockerfile?

Reply:
Having three parallel Dockefile at the same time sound complex to maintain...

Could be the same be achieved with just one Dockerfile and some kind of optionality in the FROM parameter? I mean, using the node-slim image we have now as default but being able to get a different one injected from the building system (using ARG?)

Reply:
I agree with @fgalan.

I thing the best solution is to have multiple stages on the dockerfile

`
FROM distroless/node AS distroless
....
FROM node-slim AS slim
...
FROM pm2 AS pm2
`
and on the build hook something like this:

`
docker image build --target=slim  -t ${IMAGE_NAME} ....
docker image build --target=distroless  -t ${IMAGE_NAME}-distroless ....
`

Reply:
That sounds like a plan - I'll update *this repo only* as an example.

I assume that the default will be node-slim (which must be the final one on the list)